CtrlVAF/CtrlVAF.Core/EventHandlers/EventDispatcher.cs

```diff
@@ -0,0 +1,136 @@
+using CtrlVAF.Commands.Commands;
+using CtrlVAF.Commands.Handlers;
+using CtrlVAF.Core;
+using CtrlVAF.Models;
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace CtrlVAF.Commands
+{
+
+    //    /// <summary>
+    //    /// Main command dispatcher entry method. Typical usage of this method would be inside an event handler method inside the vault application base class.
+    //    /// Once called, the dispatcher will locate any ICommandHandlers using the same TCommand interface and invoke their handle method.
+    //    /// </summary>
+    //    /// <typeparam name="TCommand">The type of command for the dispatcher to located</typeparam>
+    //    /// <param name="command">The actual command itself</param>
+    //    /// <param name="throwExceptions">Whether or not to stop executing ICommandHandlers upon exceptions and throw the exception</param>
+    //    /// <param name="exceptionHandler">An exception handler to pass along or handle any ICommandHandler exceptions</param>
+
+    public class EventDispatcher : Dispatcher
+    {
+
+        /// <summary>
+        /// Typical usage of this class would be inside an event handler method inside the vault application base class.
+        /// Once called, the dispatcher will locate any ICommandHandlers using the same TCommand interface and invoke their handle method.
+        /// </summary>
+        /// <param name="command">A command of type TCommand that has inherited from <see cref="Commands.IEventCommand{T}"/></param>
+        /// <param name="throwExceptions">Whether or not to stop executing ICommandHandlers upon exceptions and throw the exception</param>
+        /// <param name="exceptionHandler">An exception handler to pass along or handle any ICommandHandler exceptions</param>
+        public EventDispatcher()
+        {
+        }
+
+        /// <inheritdoc/>
+        public override void Dispatch(params ICtrlVAFCommand[] commands)
+        {
+            IncludeAssemblies(Assembly.GetCallingAssembly());
+
+            var concreteTypes = GetTypes(commands);
+
+            HandleConcreteTypes(concreteTypes, commands);
+        }
+
+        protected internal override IEnumerable<Type> GetTypes(params ICtrlVAFCommand[] commands)
+        {
+            // Instantiate a handlerType according to the TCommand type provided
+            Type handlerType = typeof(IEventHandler<>);
+            List<Type> dispatchableHandlerTypes = new List<Type>();
+
+            foreach (ICtrlVAFCommand command in commands)
+            {
+                Type commandType = command.GetType();
+
+                //If the concrete types have already been retrieved and cached before, simply handle those
+                if (TypeCache.TryGetValue(commandType, out var cachedTypes))
+                {
+                    dispatchableHandlerTypes.AddRange(cachedTypes.Distinct());
```

Comment:
I think the Distinct call should be moved to when handlerTypes are moved to the TypeCache on line 85.
Otherwise Distinct is called way more than necessary :)

Reply:
Oh yeah, I forgot to mention this. I was having some weird issue where the result from trygetvalue was getting me duplicates (even though it didn't contain them). But I might've gone a bit wild with the Distinct() calls. I'll look into optimizing.